### PR TITLE
Add flash attention

### PR DESCRIPTION
**Pull request by zhanghuiyao**

### CODE REVIEW

File: examples/stable_diffusion_v2/ldm/modules/attention.py [hunk 1]
```diff
@@ -142,6 +162,34 @@ def rearange_in(x):
         k = rearange_in(k)
         v = rearange_in(v)
 
+        if self.enable_flash_attention and q.shape[1] % 16 == 0 and k.shape[1] % 16 == 0:
+            out = self.flash_attention(q, k, v)
+        else:
+            out = self.custom_attention(q, k, v, mask)
+
+        def rearange_out(x):
```
Comment (DongyaoZhu) [inline examples/stable_diffusion_v2/ldm/modules/attention.py:170]: is there a reason we are having a closure here, or not having this function outside

File: examples/stable_diffusion_v2/ldm/modules/attention.py [hunk 2]
```diff
@@ -158,32 +206,56 @@ def rearange_in(x):
...
+class FlashAttention(nn.Cell):
+    def __init__(self, head_num, head_dim):
+        super(FlashAttention, self).__init__()
+        self.head_num = head_num
+        self.scale = head_dim ** -0.25
+        self.flash_attention = get_flash_attention(tiling_stgy_name='xunfei')
+        self.reshape = ops.Reshape()
+        self.scale_mul = ops.Mul()
+
+    def construct(self, q, k, v, attention_mask=None, dropout_mask=None, alibi_mask=None):
```
Comment (DongyaoZhu) [inline examples/stable_diffusion_v2/ldm/modules/attention.py:221]: might cite the paper on alibi

File: examples/stable_diffusion_v2/ldm/modules/attention.py [hunk 3]
```diff
@@ -158,32 +206,56 @@ def rearange_in(x):
...
-            return x
 
-        out = rearange_out(out)
-        return self.to_out(out)
+class FlashAttention(nn.Cell):
+    def __init__(self, head_num, head_dim):
+        super(FlashAttention, self).__init__()
+        self.head_num = head_num
+        self.scale = head_dim ** -0.25
+        self.flash_attention = get_flash_attention(tiling_stgy_name='xunfei')
```
Comment (DongyaoZhu) [inline examples/stable_diffusion_v2/ldm/modules/attention.py:217]: could you please move tiling_stgy_name to init's arg?

File: examples/stable_diffusion_v2/ldm/modules/attention.py [hunk 4]
```diff
@@ -142,6 +162,34 @@ def rearange_in(x):
         k = rearange_in(k)
         v = rearange_in(v)
 
+        if self.enable_flash_attention and q.shape[1] % 16 == 0 and k.shape[1] % 16 == 0:
```
Comment (DongyaoZhu) [inline examples/stable_diffusion_v2/ldm/modules/attention.py:165]: is there a reason this condition is not judged in init already? i.e. self.attention = XA if not enable_fa else FA?

File: examples/stable_diffusion_v2/configs/v2-train.yaml [hunk 5]
```diff
@@ -33,6 +33,7 @@ model:
         #num_heads: 8  #wukong
         num_head_channels: 64  # SD_VERSION v2.0
         use_spatial_transformer: True
+        enable_flash_attention: True
```
Comment (DongyaoZhu) [inline examples/stable_diffusion_v2/configs/v2-train.yaml:36]: what about other configs, v1, inference, etc.

**Review by DongyaoZhu — COMMENTED** [review]
if please address the comments

File: examples/stable_diffusion_v2/ldm/modules/attention.py (hunk 1 above)
Comment (zhanghuiyao) [reply examples/stable_diffusion_v2/ldm/modules/attention.py:170]: > is there a reason we are having a closure here, or not having this function outside

Just for the uniformity of the code style, I can expand this ‘closure’ if necessary

File: examples/stable_diffusion_v2/ldm/modules/attention.py (hunk 2 above)
Comment (zhanghuiyao) [reply examples/stable_diffusion_v2/ldm/modules/attention.py:221]: This parameter is currently not used, and it is planned to delete it

File: examples/stable_diffusion_v2/ldm/modules/attention.py (hunk 3 above)
Comment (zhanghuiyao) [reply examples/stable_diffusion_v2/ldm/modules/attention.py:217]: This is an internal parameter, and users are not expected to modify it.

File: examples/stable_diffusion_v2/ldm/modules/attention.py (hunk 4 above)
Comment (zhanghuiyao) [reply examples/stable_diffusion_v2/ldm/modules/attention.py:165]: Here is a dynamic judgment based on the tensor of the actual operation

File: examples/stable_diffusion_v2/configs/v2-train.yaml (hunk 5 above)
Comment (zhanghuiyao) [reply examples/stable_diffusion_v2/configs/v2-train.yaml:36]: Only this config file is currently verified

File: examples/stable_diffusion_v2/ldm/modules/attention.py [hunk 6]
```diff
@@ -21,6 +21,17 @@
...
+import mindspore.numpy as mnp
+from mindspore import Tensor
+
+try:
+    from mindspore.ops._op_impl._custom_op.flash_attention.flash_attention_impl import get_flash_attention
+    FLASH_IS_AVAILABLE = True
+    print("flash_attention is available.")
+except:
+    FLASH_IS_AVAILABLE = False
+    print("no module 'flash_attention'. Processing without...")
```
[the remark this answers is not in the document]
Comment (zhanghuiyao) [reply examples/stable_diffusion_v2/ldm/modules/attention.py:33]: may I should change it to "flash attention is unavailable."

File: examples/stable_diffusion_v2/ldm/modules/attention.py [hunk 7]
```diff
@@ -121,6 +132,15 @@ def __init__(self, query_dim, context_dim=None, heads=8, dim_head=64, dropout=1.
             nn.Dropout(dropout) if is_old_ms_version() else nn.Dropout(p=1 - dropout),
         )
 
+        self.enable_flash_attention = enable_flash_attention and \
+                                      FLASH_IS_AVAILABLE and \
+                                      (ms.context.get_context("device_target") == "Ascend")
+        if enable_flash_attention and not self.enable_flash_attention:
+            print("WARNING: flash attention not available.")
+
+        self.custom_attention = CustomAttention(dim_head)
```
[the remark this answers is not in the document]
Comment (zhanghuiyao) [reply examples/stable_diffusion_v2/ldm/modules/attention.py:141]: ok

**Review by DongyaoZhu — APPROVED** [review]
LGTM

File: examples/stable_diffusion_v2/ldm/modules/attention.py [hunk 8]
```diff
@@ -158,32 +205,58 @@ def rearange_in(x):
...
 
-        out = rearange_out(out)
-        return self.to_out(out)
+# reference to https://arxiv.org/abs/2205.14135
+class FlashAttention(nn.Cell):
+    def __init__(self, head_num, head_dim):
+        super(FlashAttention, self).__init__()
+        self.head_num = head_num
+        self.scale = head_dim ** -0.25
+        self.flash_attention = get_flash_attention(tiling_stgy_name='xunfei')
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/attention.py:217]: `tiling_stgy_name='xunfei'` 

I think it is a little weird to name the strategy with a company name.  Maybe name by this tiling algorithm?

Comment (zhanghuiyao) [reply examples/stable_diffusion_v2/ldm/modules/attention.py:217]: This is a fixed parameter of the internal operator of the current mindspore version, and the latest code on the mindspore master/r2.0/r2.1 branch will be modified to "tiling_stgy_name='sparse'"

**Comment by SamitHuang** [comment]
btw, please run pre-commit and fix the format issues

**Pull request merged**